### PR TITLE
fix(claude): respect anthropic-ratelimit-unified-reset on 429

### PR DESCRIPTION
**Pull request by ak2k**
## Summary

- On Anthropic 429s, parse `anthropic-ratelimit-unified-reset` (epoch
  seconds) and `Retry-After` from response headers and pipe through
  `statusErr.retryAfter` so the auth conductor schedules a correctly-sized
  cooldown instead of falling back to the 30m exponential cap.
- Mirrors the sibling pattern at `parseCodexRetryAfter` in
  `internal/runtime/executor/codex_executor.go:964` — same
  `(headers, now) → *time.Duration` signature, same past-reset → nil
  discipline.
- Cooldowns are clamped at 7 days (Anthropic's longest documented quota
  window) so a malformed/buggy upstream value can't lock out a credential
  permanently.

## Why

The Claude executor returned bare `statusErr{code, msg}` on errors,
leaving `result.RetryAfter` unset on 429s. The conductor then fell back
to `nextQuotaCooldown`'s exponential ladder, capped at 30 min via
`quotaBackoffMax` (`sdk/cliproxy/auth/conductor.go:73`). Anthropic's
quota windows are 5h and 7d — far longer than 30 min — so with
`routing.strategy: fill-first` an exhausted credential re-enters
selection every 30 min, immediately 429s, and loops until the actual
window resets.

Real captured 429 from `api.anthropic.com/v1/messages`:

    status: 429
    anthropic-ratelimit-unified-status: rejected
    anthropic-ratelimit-unified-representative-claim: five_hour
    anthropic-ratelimit-unified-reset: 1777500000   # epoch sec, ~5h after request
    anthropic-ratelimit-unified-7d-status: allowed
    anthropic-ratelimit-unified-7d-reset: 1777561200

## Validation

    go test -race -count=1 -run 'ParseClaude|NewClaudeStatusErr' \
      ./internal/runtime/executor/...
    go vet ./internal/runtime/executor/...
    go build -o test-output ./cmd/server && rm test-output

13 unit tests cover: epoch parsing, clamp behavior (year 5138 → 7d cap,
absurd seconds → 7d cap, 10y HTTP-date → 7d cap), Retry-After
seconds/HTTP-date fallback, past-reset → nil, only-on-429 gate, and
header priority. No `internal/translator` or `AGENTS.md` paths touched.
Falls back to the existing exponential cooldown when neither header is
present; `disable-cooling: true` continues to bypass.

### CODE REVIEW

File: internal/runtime/executor/claude_retry_after.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+// malformed upstream value (e.g. an off-by-1000 ms-vs-s mistake producing a
+// year-5138 epoch) would otherwise lock out a credential effectively forever.
+const maxClaudeRetryAfter = 7 * 24 * time.Hour
+
+// newClaudeStatusErr builds a statusErr for an Anthropic Claude error
+// response, populating retryAfter on 429 from the response headers.
+func newClaudeStatusErr(statusCode int, headers http.Header, body []byte) statusErr {
+	err := statusErr{code: statusCode, msg: string(body)}
+	if statusCode == http.StatusTooManyRequests {
+		if d := parseClaudeRetryAfter(headers, time.Now()); d != nil {
```
[the remark this answers is not in the document]
Comment (ak2k) [reply internal/runtime/executor/claude_retry_after.go]: Good catch — accepted. Anchoring `now` to the response's `Date` header in `parseClaudeRetryAfter` (commit f759f89). The codex sibling can rely on `time.Now()` because it parses `error.resets_in_seconds` (relative), but the unified-reset path is an absolute epoch and is genuinely clock-skew sensitive. Falls back to the caller-supplied wall-clock when `Date` is missing or unparseable. Added `TestParseClaudeRetryAfterAnchorsToServerDate` to cover the skew case (1h-ahead client clock + a 5-min reset — without anchoring it would return nil; with anchoring it returns 5min).

File: internal/runtime/executor/claude_retry_after.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+func parseClaudeRetryAfter(headers http.Header, now time.Time) *time.Duration {
+	if v := strings.TrimSpace(headers.Get("anthropic-ratelimit-unified-reset")); v != "" {
+		if epoch, errParse := strconv.ParseInt(v, 10, 64); errParse == nil {
+			if d := clampedRetryAfter(time.Unix(epoch, 0).Sub(now)); d != nil {
+				return d
+			}
+		}
+	}
+	if v := strings.TrimSpace(headers.Get("Retry-After")); v != "" {
+		if secs, errParse := strconv.Atoi(v); errParse == nil && secs > 0 {
```
[the remark this answers is not in the document]
Comment (ak2k) [reply internal/runtime/executor/claude_retry_after.go]: Both points accepted (commit f759f89).

**`Atoi` → `ParseInt(_, 10, 64)`:** real concern even though most builds are 64-bit — switching is free and removes a 32-bit footgun. New `Retry-After exceeds 32-bit int — parsed as int64, clamped to max` test exercises a `4294967296` value (> `math.MaxInt32`) which would have failed under `Atoi` on 32-bit GOARCH; under `ParseInt` it parses cleanly and the existing pre-multiplication clamp keeps it bounded at the 7d cap.

**`secs >= 0`:** agreed — RFC 9110 §10.2.3 defines `Retry-After: 0` as "retry immediately" and propagating `*time.Duration(0)` is the right way to convey that to the conductor (which evaluates `now.Add(*RetryAfter)` and the credential becomes ready on the next poll). Updated the existing `Retry-After zero` test to expect `durPtr(0)` and added a `Retry-After negative` case to confirm the previous "nil for negative" behavior is preserved (negative parses fine, fails the `>= 0` gate, falls through to `http.ParseTime`, which fails on a negative integer, returns nil).

File: internal/runtime/executor/claude_retry_after.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+			}
+		}
+	}
+	return nil
+}
+
+// clampedRetryAfter returns nil for non-positive durations and caps positive
+// durations at maxClaudeRetryAfter.
+func clampedRetryAfter(d time.Duration) *time.Duration {
+	if d <= 0 {
```
[the remark this answers is not in the document]
Comment (ak2k) [reply internal/runtime/executor/claude_retry_after.go]: Accepted (commit f759f89). Now `d < 0` returns nil and `d == 0` returns `&time.Duration(0)`, consistent with the seconds-path change above. One subtle behavior change worth calling out: when `unified-reset` resolves to *exactly* the server's `Date` (d == 0), the function now short-circuits with `&Duration(0)` instead of falling through to `Retry-After`. Past-reset fall-through is preserved (d < 0 still returns nil — covered by the existing `unified-reset past epoch — falls through to Retry-After` test). Updated the doc-comment on `clampedRetryAfter` to spell this out.

**Review by luispater — APPROVED** [review]
## Summary
This PR teaches the Claude executor to respect Anthropic’s rate-limit reset hints on HTTP 429 by parsing `anthropic-ratelimit-unified-reset` (epoch seconds) with a `Retry-After` fallback, then piping the result through `statusErr.retryAfter` so the auth conductor can schedule an appropriate cooldown (instead of the default 30m exponential cap). Cooldowns are clamped to 7 days, and calculations are anchored to the response `Date` header when present to reduce client clock-skew impact.

## Key findings
- Good scope: changes are localized to the Claude executor plus a helper in `internal/runtime/executor/helps/`.
- Correct propagation path: `statusErr.retryAfter` is picked up by the conductor via `RetryAfter()` and used to set `NextRetryAfter` for 429s.
- Solid tests: covers epoch parsing, clamp behavior, `Retry-After` (seconds + HTTP-date), negative handling, and `Date` anchoring.

## Suggestions (non-blocking)
- Consider whether `Retry-After: 0` should translate to an immediate retry in our scheduler/selector behavior, or whether a minimal floor (or leaving it as “no hint”) is preferable for stability under upstream anomalies.
- With `Date` anchoring, a 1-second ordering mismatch between `Date` and epoch reset could cause a “past reset → nil” fallthrough; consider whether a small tolerance is desirable.

## Test plan
- `go test ./internal/runtime/executor/...`
- `go build -o test-output ./cmd/server && rm test-output`

This is an automated Codex review result and still requires manual verification by a human reviewer.

**Review by luispater — APPROVED** [review]
## Summary

This PR adds proper parsing of Anthropic's `anthropic-ratelimit-unified-reset` header on 429 responses, piping the parsed duration through `statusErr.retryAfter` so the auth conductor can schedule correctly-sized cooldowns instead of falling back to the 30-minute exponential cap.

## Key findings

- No blocking issues. Implementation mirrors the established Codex pattern with improvements (Date anchoring, int64 parsing, 7-day clamp).
- 13 table-driven tests cover edge cases comprehensively.
- CI passes. No merge conflicts.

## Non-blocking

- Two commits share identical messages -- differentiating would improve git-log readability.
- `Retry-After: 0` semantic divergence from `parseCodexRetryAfter` is safe but worth a code comment.

## Verdict

Approve.

---

> This is an automated Claude Code review result and still requires manual verification by a human reviewer.

**Comment by ak2k** [comment]
Force-pushed a rebase onto current `dev` (was `b7c4967`, now `764d391`) — the branch had gone `mergeable: false` after the executor refactor.

`claude_executor.go` was split into `claude_executor_{execute,stream,tokens}.go`, which moved all six `statusErr` construction sites this PR touches; the old diff rejected 6 of its 7 hunks. The six sites are relocated verbatim into their new files — no behavior change from the version you last saw. `newClaudeStatusErr` and the `helps` package are unchanged.

Also squashed to a single commit. `go build`, `go vet`, `gofmt` and the full `go test ./...` pass on the rebased branch.
